### PR TITLE
Add type annotations to optix.

### PR DESCRIPTION
**Pull request by aslanides**
@mtthss to review.

### CODE REVIEW

**Comment by aslanides** [comment]
Relates to issue #1555.

**Review by gnecula — APPROVED** [review]
Thank you for adding type annotations. 

LGTM, but I am wondering if all the concrete type annotations (`float`, `jnp.ndarray`) still hold if this code
is under jax.jit?

**Comment by aslanides** [comment]
The float annotations only pertain to the 'constructor' functions which return the InitUpdate pairs. All other annotations are jnp.ndarrays or Any; jnp.ndarray inputs/output types are invariant under jit, if I undertand correctly.

**Comment by gnecula** [comment]
Sounds good. Should I merge this or should we wait for @mtthss to review?

File: jax/experimental/optix.py
Comment (tomhennigan) [inline jax/experimental/optix.py:471]: Should we have `AdamState` (etc) for the common cases (e.g. `AdamState = Sequence[ScaleByAdamState, ScaleState, ScaleState]`).

**Comment by aslanides** [comment]
> Sounds good. Should I merge this or should we wait for @mtthss to review?

Let's wait for @mtthss to take a look.

File: jax/experimental/optix.py
Comment (aslanides) [reply jax/experimental/optix.py:471]: Sure, I'll add that now.

Comment (aslanides) [reply jax/experimental/optix.py:471]: I'll leave that for a follow-up. I think we should improve the other types before we do that -- e.g. by making InitUpdate parametric/generic w.r.t. the state, so that we can have InitUpdate[Adam] rather than shoving unions of things into InitUpdate. WDYT?

Comment (tomhennigan) [reply jax/experimental/optix.py:471]: Sure, I think that will require quite a bit of refactoring though (e.g. I'm pretty sure I've tried to subclass NamedTuple and Generic before and it failed because of a metaclass conflict)? Up to you if you think its worth it.

**Comment by mtthss** [comment]
Looks great! Thank you very much!

**Comment by mtthss** [comment]
LGTM

File: jax/experimental/optix.py
Comment (aslanides) [reply jax/experimental/optix.py:471]: Ack -- will return to this at some point if I have time.

**Comment by aslanides** [comment]
@gnecula this can go in now -- thanks all :)

**Pull request merged**
